src/java.base/share/classes/java/lang/StringLatin1.java

```diff
@@ -388,30 +388,39 @@ public static boolean regionMatchesCI(byte[] value, int toffset,
                                           byte[] other, int ooffset, int len) {
         int last = toffset + len;
         while (toffset < last) {
-            char c1 = (char)(value[toffset++] & 0xff);
-            char c2 = (char)(other[ooffset++] & 0xff);
-            if (c1 == c2) {
-                continue;
-            }
-            int u1 = CharacterDataLatin1.instance.toUpperCase(c1);
-            int u2 = CharacterDataLatin1.instance.toUpperCase(c2);
-            if (u1 == u2) {
+            byte b1 = value[toffset++];
+            byte b2 = other[ooffset++];
+            if (b1 == b2 || latin1EqualsIgnoreCase(b1, b2)) {
                 continue;
             }
             return false;
         }
         return true;
     }
-
     public static boolean regionMatchesCI_UTF16(byte[] value, int toffset,
                                                 byte[] other, int ooffset, int len) {
         int last = toffset + len;
         while (toffset < last) {
             char c1 = (char)(value[toffset++] & 0xff);
             char c2 = StringUTF16.getChar(other, ooffset++);
+            // Fast path when codepoints are the same
             if (c1 == c2) {
                 continue;
             }
+            // Fast path for latin1 letters with folding case
+            if (StringUTF16.latin1EqualsIgnoreCase(c1, c2)) {
+                continue;
+            }
+            // Fast path if both code points are latin1
+            if (c2 <= 0XFF) {
+                return false;
+            }
+            // Fast path for unicode codepoints which cannot fold info latin1
+            if (c2 > 0XFF && !CharacterData.foldsToLatin1(c2)) {
```

Comment:
```suggestion
            else if (!CharacterData.foldsToLatin1(c2)) {
```

Or fold the checks into a single if (c2 <= 0xFF || !CharacterData.foldsToLatin1(c2))

Reply:
This code is outdated and was possibly also incorrect. I did not push my latest changes yesterday and got sidetracked today by #12623. 

There are various code paths in this PR which would benefit from a general speedup in Character.toUpperCase, Character.toUpperCase.